bluemira/geometry/wire.py

```diff
@@ -153,6 +152,44 @@ def discretize(
             points = cadapi.discretize(self.shape, ndiscr=ndiscr, dl=dl)
         return Coordinates(points.T)
 
+    def discretize_reduced(
+        self, ndiscr: int = 100, byedges: bool = False, dl: Optional[float] = None
+    ) -> Coordinates:
+        """
+        Discretize the wire in ndiscr equidistant points or with a reference dl
+        segment step according to discretize function. Will then reduce the number
+        of discretizations along straight segments to reduce computational needs.
+
+        Parameters
+        ----------
+        ndiscr:
+            Number of points to discretize to
+        byedges:
+            Whether or not to discretise by edges. If True, each edge is
+            discretized separately using an approximated distance
+            (wire.Length/ndiscr) or the specified dl. If True, it is
+            possible that ndiscr is larger than specified.
+        dl:
+            Discretise by length, overriding ndiscr
+
+        Returns
+        -------
+        Coordinates of the discretized points.
+        """
+        points = self.discretize(ndiscr, byedges, dl)
+        points_copy = points
+        for i in range(1, len(points) - 1):
+            # go in reverse so element deletion doesn't change positional argument
+            j = len(points) - 1 - i
+            p0 = points[:, j + 1]
+            p1 = points[:, j]
+            p2 = points[:, j - 1]
+            l1 = (p1 - p0) / np.linalg.norm(p1 - p0)
+            l2 = (p2 - p1) / np.linalg.norm(p2 - p1)
+            if np.equal(np.dot(l1, l2), 1.0):
+                points_copy = np.delete(points_copy, j, 1)
```

Comment:
can we vectorise this. The steps would be:

1) get all l1 and l2 as a 2d array something like (I think my indexes are wrong on the points)
```
p0 = points[:, 2:-2]
p1 = points[:, 1: -3]
p2 = points[:,  : -4]
l1 = p1 -p0/ np.linalg.norm(p1 - p0, axis=1)
l2 = p2 -p1/ np.linalg.norm(p2 - p1, axis=1)
```
2) `mask = np.ones_like(points, dtype=bool)`
3) `mask[:, np.equal(np.dot(..., axis=..), 1.0)] = False`
2) `return Coordinates(points[mask])`

Reply:
Made these changes, the dot product came out quite messy not sure if there is a better way to do it than I have...

```
diag_dot = np.around(
                np.append(np.append(0.0, np.diag(np.inner(l1, l2))), 0.0)
            ).astype(int)
```

numpy dot (or other versions of it) don't take dot products for each row and put into a new array of size N but just performs matrix multiplication. Inner was the closest I could get but it does all the combinations, hence the diagonalisation. The rounding and changing type to integer was because otherwise the numpy equals comparison gave wrong answers.